pkg/clustermanager/cluster_manager.go

```diff
@@ -480,105 +479,12 @@ func (c *ClusterManager) EKSAClusterSpecChanged(ctx context.Context, cluster *ty
-		if err != nil {
-			return false, err
-		}
-		vdc := datacenterConfig.(*v1alpha1.VSphereDatacenterConfig)
-		if !reflect.DeepEqual(existingVdc.Spec, vdc.Spec) {
-			logger.V(3).Info("New provider spec is different from the new spec")
-			return true, nil
-		}
-
-		for _, config := range machineConfigs {
-			mc := config.(*v1alpha1.VSphereMachineConfig)
-			machineConfigMap[mc.Name] = mc
-		}
-		existingCpVmc, err := c.clusterClient.GetEksaVSphereMachineConfig(ctx, cc.Spec.ControlPlaneConfiguration.MachineGroupRef.Name, cluster.KubeconfigFile, newClusterSpec.Cluster.Namespace)
-		if err != nil {
-			return false, err
-		}
-		cpVmc := machineConfigMap[newClusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name]
-		if !reflect.DeepEqual(existingCpVmc.Spec, cpVmc.Spec) {
-			logger.V(3).Info("New control plane machine config spec is different from the existing spec")
-			return true, nil
-		}
-		for _, workerNodeGroupConfiguration := range cc.Spec.WorkerNodeGroupConfigurations {
-			existingWnVmc, err := c.clusterClient.GetEksaVSphereMachineConfig(ctx, workerNodeGroupConfiguration.MachineGroupRef.Name, cluster.KubeconfigFile, newClusterSpec.Cluster.Namespace)
-			if err != nil {
-				return false, err
-			}
-			wnVmc := machineConfigMap[workerNodeGroupConfiguration.MachineGroupRef.Name]
-			if !reflect.DeepEqual(existingWnVmc.Spec, wnVmc.Spec) {
-				logger.V(3).Info("New worker node machine config spec is different from the existing spec")
-				return true, nil
-			}
-		}
-		if cc.Spec.ExternalEtcdConfiguration != nil {
-			existingEtcdVmc, err := c.clusterClient.GetEksaVSphereMachineConfig(ctx, cc.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name, cluster.KubeconfigFile, newClusterSpec.Cluster.Namespace)
-			if err != nil {
-				return false, err
-			}
-			etcdVmc := machineConfigMap[newClusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name]
-			if !reflect.DeepEqual(existingEtcdVmc.Spec, etcdVmc.Spec) {
-				logger.V(3).Info("New etcd machine config spec is different from the existing spec")
-				return true, nil
-			}
-		}
-	case v1alpha1.CloudStackDatacenterKind:
-		existingCsdc, err := c.clusterClient.GetEksaCloudStackDatacenterConfig(ctx, cc.Spec.DatacenterRef.Name, cluster.KubeconfigFile, newClusterSpec.Cluster.Namespace)
-		if err != nil {
-			return false, err
-		}
-		csDc := datacenterConfig.(*v1alpha1.CloudStackDatacenterConfig)
-		if !existingCsdc.Spec.Equal(&csDc.Spec) {
-			logger.V(3).Info("New provider spec is different from the new spec")
-			return true, nil
-		}
-
-		machineConfigsSpecChanged, err := c.cloudstackMachineConfigsSpecChanged(ctx, cc, cluster, newClusterSpec, machineConfigs)
-		if err != nil {
-			return false, err
-		}
-		if machineConfigsSpecChanged {
-			return true, nil
-		}
-		// TODO: Make sure happy paths at least are unit tested, especially if moving code to provider
+	case v1alpha1.VSphereDatacenterKind, v1alpha1.CloudStackDatacenterKind:
```

Comment:
Since you implemented it for the two providers (btw thanks!), I would vote to use this directly from the upgrade workflow and avoid injecting the provider here.

In facts providers already implement a method `UpgradeNeeded` that is being called from the workflow. I would suggest moving the implementation of `SpecChanged` to `UpgradeNeeded`